appsmithorg/appsmith
# fix: remove debounce from RTE component

## Description

Remove debounce from RTE component

It was noticed that `onEditorChange` was called also when RTE component was focused, unfocused ( cases where value was not changed, instead editor's value changes. ) in such cases, we don't want to call `onValueChange`. 
Hence, we added logic to only update value when we have an actual change in value.

Fixes following issues:
#14045
#7231 
#10419 
https://github.com/appsmithorg/appsmith/issues/14708
#14922 

This PR also fixes the issues that were caught during the development process:

* If Input type is markdown or html and when you cut the text inside the RTE widget then text re-appears (in this case default text is none). Here is the video that demonstrates this behaviour: 

https://user-images.githubusercontent.com/14138515/174781044-4ce4f760-1e3c-4f1b-a195-c876a0ddd797.mov

* The isDirty property of the widget was not working as expected. isDirty property was not getting set to false when the default text is changed. 
* ResetWidget functionality of the widget was not working properly. It was observed that for resetting the widget, `resetWidget` function was needed to be called twice. 
* isDirty was not setting to false for the last step in the following scenario: Update normal text(isDirt true) -> Update Default Text(isDirt false) -> Reset widget(isDirt false) ->. Update default Text(isDirt true)

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

This PR adds the following cypress test cases into the file `RichTextEditor_spec.js`:

* Updated the`Check isDirty meta property` test case to check for the following scenario - `Update normal text -> isDirty True -> Update Default Text -> isDirty false -> Reset widget ->. Update default Text -> isDirty false`
* Check if the binding is getting removed from the text and the RTE widget Once the default text is cleared.
* Check if text does not re-appear when cut, inside the RTE widget
* Check if the cursor position is at the end for the RTE widget
* Check if different font size texts are supported inside the RTE widget

Following scenario needs to be tested manually: 
* When Input Type is Markdown - cursor jumps(Additional scenario for #7231 )
* The google sheet issue mentioned in the github issue #14045 needs to be tested manually.

## Checklist:





## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>


    // Code coverage diff between base branch:release and head branch: fix/RTE-remove-debounce 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 56.71 **(0)** | 38.66 **(-0.01)** | 36.22 **(0)** | 56.96 **(0)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**</details>
